File: commons/tools/src/main/java/cn/wjybxx/common/tools/protobuf/PBElement.java
/*
 * Copyright 2023 wjybxx(845740757@qq.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package cn.wjybxx.common.tools.protobuf;

import cn.wjybxx.common.tools.util.Line;
import cn.wjybxx.common.tools.util.Utils;

import javax.annotation.Nonnull;
import java.util.*;

/**
 * protobuf所有元素抽象
 *
 * @author wjybxx
 * date - 2023/10/7
 */
public abstract class PBElement {

    /** 简单名 */
    private String simpleName;
    /** 注释 -- 包含注解的原始注释 */
    private final List<String> comments = new ArrayList<>();
    /** 可选项 */
    private final Map<String, String> options = new LinkedHashMap<>();

    /** 定义该元素的元素 */
    private PBElement enclosingElement;
    /** 嵌套定义的元素 -- 任何便捷查询都是筛选后的快照 */
    private final List<PBElement> enclosedElements = new ArrayList<>();
    /** 注解数据 -- 同一类型允许重复 */
    private final List<PBAnnotation> annotations = new ArrayList<>();

    /** 定义元素的源代码行 -- 可能为null，表示非源码文件定义 */
    private Line sourceLine;
    /** 结束元素定义的行 -- 可能为null，特殊用途（节选） */
    private Line sourceEndLine;

    // region

    @Nonnull
    public abstract PBElementKind getKind();

    public PBElement addEnclosedElement(PBElement enclosed) {
        Objects.requireNonNull(enclosed);
        if (this == enclosed) throw new IllegalArgumentException("add self");

        enclosed.enclosingElement = this;
        enclosedElements.add(enclosed);
        return this;
    }

    public PBElement addAnnotation(PBAnnotation annotation) {
        annotations.add(annotation);
        return this;
    }

    public PBElement addComment(String comment) {
        comments.add(comment);
        return this;
    }

    public PBElement addOption(String key, String value) {
        options.put(key, value);
        return this;
    }

    /** 获取指定类型注解 */
    public PBAnnotation getAnnotation(String type) {
        return annotations.stream()
                .filter(e -> e.type.equals(type))
                .findFirst()
                .orElse(null);
    }

    /** 获取指定类型注解 */
    public List<PBAnnotation> getAnnotations(String type) {
        return annotations.stream()
                .filter(e -> e.type.equals(type))
                .toList();
    }

    /** 获取可选项的值 */
    public String getOption(String name) {
        return options.get(name);
    }

    /** 获取去除掉双引号的可选项值 */
    public String getUnquoteOption(String name) {
        return Utils.unquote(options.get(name));
    }

    // endregion

    // region getter/setter

    public String getSimpleName() {
        return simpleName;
    }

    public PBElement setSimpleName(String simpleName) {
        this.simpleName = simpleName;
        return this;
    }

    public Line getSourceLine() {
        return sourceLine;
    }

    public PBElement setSourceLine(Line sourceLine) {
        this.sourceLine = sourceLine;
        return this;
    }

    public Line getSourceEndLine() {
        return sourceEndLine;
    }

    public PBElement setSourceEndLine(Line sourceEndLine) {
        this.sourceEndLine = sourceEndLine;
        return this;
    }

    public List<String> getComments() {
        return comments;
    }

    public PBElement getEnclosingElement() {
        return enclosingElement;
    }

    public Map<String, String> getOptions() {
        return options;
    }

    public List<PBElement> getEnclosedElements() {
        return enclosedElements;
    }

    public List<PBAnnotation> getAnnotations() {
        return annotations;
    }

    // endregion

    @Override
    public final String toString() {
        StringBuilder stringBuilder = new StringBuilder()
                .append(getKind()).append("{")
                .append("simpleName='").append(simpleName).append('\'')
                .append(", enclosingElement=").append(enclosingElement == null ? null : enclosingElement.getSimpleName());

        toString(stringBuilder);
        return stringBuilder
                .append('}').toString();
    }

    protected void toString(StringBuilder sb) {

    }

}

File: csharp/Wjybxx.BigCatTool.Core/src/Protobuf/PBElement.cs
#region LICENSE

// Copyright 2025 wjybxx(845740757@qq.com)
// 
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// 
//     http://www.apache.org/licenses/LICENSE-2.0
// 
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#endregion

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Wjybxx.BigCatTool.Core;
using Wjybxx.Commons;
using Wjybxx.Commons.Collections;

namespace Wjybxx.BigCatTool.Protobuf
{
/// <summary>
/// PB元素
/// </summary>
public abstract class PBElement
{
    private static readonly IDictionary<string, string> EMPTY_OPTIONS = ImmutableDictionary<string, string>.Empty;
#nullable disable
    /** 简单名 */
    private string simpleName;
    /** 定义该元素的元素 */
    private PBElement enclosingElement;
    /** 嵌套定义的元素 -- 任何便捷查询都是筛选后的快照 */
    private readonly List<PBElement> enclosedElements = new();

    /** 注释 -- 包含注解的原始注释 */
    private readonly List<string> comments = new();
    /** 注解数据 -- 同一类型允许重复；可能有解析器动态附加的数据 */
    private readonly List<Annotation> annotations = new();
    /** 可选项 -- 数据脚本中通常不使用该机制，而是使用注解；只建议文件使用options；延迟初始化以减少开销 */
    private IDictionary<string, string> options = EMPTY_OPTIONS;

    /** 定义元素的开始行号 -- -1表示非源码文件定义 */
    private int startLine = -1;
    /** 定义元素的结束行号 -- -1表示非源码文件定义 */
    private int endLine = -1;
#nullable restore

    #region logic

    /// <summary>
    /// 元素的类型
    /// </summary>
    public abstract PBElementKind Kind { get; }

    public PBElement AddEnclosedElement(PBElement enclosed) {
        if (enclosed == null) throw new ArgumentNullException(nameof(enclosed));
        if (this == enclosed) throw new ArgumentException("this == enclosed");

        enclosed.enclosingElement = this;
        enclosedElements.Add(enclosed);
        return this;
    }

    public PBElement AddComment(string comment) {
        comments.Add(comment);
        return this;
    }

    public PBElement AddAnnotation(Annotation annotation) {
        annotations.Add(annotation);
        return this;
    }

    public PBElement AddOption(string key, string value) {
        if (ReferenceEquals(options, EMPTY_OPTIONS)) {
            options = new Dictionary<string, string>(4);
        }
        options[key] = value;
        return this;
    }

    /** 获取指定类型注解 */
    public Annotation? GetAnnotation(string type) {
        return annotations.FirstOrDefault(e => e.type == type);
    }

    /** 获取指定类型所有注解 */
    public List<Annotation> GetAnnotations(string type) {
        return annotations.Where(e => e.type == type).ToList();
    }

    /** 获取可选项的值 */
    public string? GetOption(string name) {
        options.TryGetValue(name, out string? r);
        return r;
    }

    #endregion

#nullable disable

    #region props

    public string SimpleName {
        get => simpleName;
        set => simpleName = value;
    }

    public PBElement EnclosingElement => enclosingElement;
    public List<PBElement> EnclosedElements => enclosedElements;
    public List<string> Comments => comments;
    public List<Annotation> Annotations => annotations;
    public IDictionary<string, string> Options => options;

    public int StartLine {
        get => startLine;
        set => startLine = value;
    }
    public int EndLine {
        get => endLine;
        set => endLine = value;
    }

    #endregion

    public sealed override string ToString() {
        StringBuilder stringBuilder = new StringBuilder()
            .Append(EnumUtil.GetName(Kind)).Append("{")
            .Append("simpleName='").Append(simpleName).Append('\'')
            .Append(", enclosingElement=").Append(enclosingElement == null ? null : enclosingElement.simpleName);

        ToString(stringBuilder);
        return stringBuilder
            .Append(", startLn='").Append(startLine)
            .Append(", endLn='").Append(endLine)
            .Append('}').ToString();
    }

    protected virtual void ToString(StringBuilder sb) {
    }
}
}